File: problems/two_sum/two_sum_test.py
"""Test cases for the two_sum function."""

import pytest

from two_sum import twoSum


@pytest.mark.parametrize(
    "nums, target, expected",
    [
        ([2, 7, 11, 15], 9, [0, 1]),  # Basic case
        ([3, 2, 4], 6, [1, 2]),  # Middle elements
        ([3, 3], 6, [0, 1]),  # Duplicate elements
        ([-3, 1, 2, 3], 0, [0, 3]),  # Negative and positive mix
        ([0, 1], -1, [-1, -1]),  # Edge case: no solution
        ([1, 2, 3, 4, 5], 8, [2, 4]),  # Target at end
        ([5, 75, 25], 100, [1, 2]),  # Large numbers
        ([-1, -2, -3, -4, -5], -8, [2, 4]),  # All negative numbers
        ([0, 4, 3, 0], 0, [0, 3]),  # Duplicate zeros
        ([1, 1, 1, 1], 2, [0, 1]),  # Multiple duplicates
        ([1], 2, [-1, -1]),  # Single element, no solution
        ([2, 2], 4, [0, 1]),  # Two identical elements that sum to target
        ([1000, -1000], 0, [0, 1]),  # Large positive and negative
        ([1, 2, 3, 4, 5, 6], 11, [4, 5]),  # Larger array
        ([0, 0], 0, [0, 1]),  # Two zeros sum to zero
    ],
    ids=[
        "basic_case",
        "middle_elements",
        "duplicate_elements",
        "negative_positive_mix",
        "no_solution",
        "target_at_end",
        "large_numbers",
        "all_negative",
        "duplicate_zeros",
        "multiple_duplicates",
        "single_element",
        "identical_sum_to_target",
        "large_pos_neg",
        "larger_array",
        "two_zeros_sum_to_zero",
    ],
)
def test_two_sum(nums, target, expected):
    assert twoSum(nums, target) == expected

File: problems/two_sum/two_sum_test.cc
#include "two_sum.h"

#include <gtest/gtest.h>
#include <vector>

struct TwoSumCase {
  const std::string test_name;
  const std::vector<int> nums;
  const int target;
  const std::vector<int> expected;
};

using TwoSumTest = ::testing::TestWithParam<TwoSumCase>;

TEST_P(TwoSumTest, TestCases) {
  const TwoSumCase &test_case = GetParam();
  const std::vector<int> result = twoSum(test_case.nums, test_case.target);
  EXPECT_EQ(result, test_case.expected);
}

INSTANTIATE_TEST_SUITE_P(
    TwoSumTestCases, TwoSumTest,
    ::testing::Values(TwoSumCase{.test_name = "BasicCase",
                                 .nums = {2, 7, 11, 15},
                                 .target = 9,
                                 .expected = {0, 1}},
                      TwoSumCase{.test_name = "MiddleElements",
                                 .nums = {3, 2, 4},
                                 .target = 6,
                                 .expected = {1, 2}},
                      TwoSumCase{.test_name = "DuplicateElements",
                                 .nums = {3, 3},
                                 .target = 6,
                                 .expected = {0, 1}},
                      TwoSumCase{.test_name = "NegativePositiveMix",
                                 .nums = {-3, 1, 2, 3},
                                 .target = 0,
                                 .expected = {0, 3}},
                      TwoSumCase{.test_name = "NoSolution",
                                 .nums = {0, 1},
                                 .target = -1,
                                 .expected = {-1, -1}},
                      TwoSumCase{.test_name = "TargetAtEnd",
                                 .nums = {1, 2, 3, 4, 5},
                                 .target = 8,
                                 .expected = {2, 4}},
                      TwoSumCase{.test_name = "LargeNumbers",
                                 .nums = {5, 75, 25},
                                 .target = 100,
                                 .expected = {1, 2}},
                      TwoSumCase{.test_name = "AllNegativeNumbers",
                                 .nums = {-1, -2, -3, -4, -5},
                                 .target = -8,
                                 .expected = {2, 4}},
                      TwoSumCase{.test_name = "DuplicateZeros",
                                 .nums = {0, 4, 3, 0},
                                 .target = 0,
                                 .expected = {0, 3}},
                      TwoSumCase{.test_name = "MultipleDuplicates",
                                 .nums = {1, 1, 1, 1},
                                 .target = 2,
                                 .expected = {0, 1}},
                      TwoSumCase{.test_name = "SingleElementNoSolution",
                                 .nums = {1},
                                 .target = 2,
                                 .expected = {-1, -1}},
                      TwoSumCase{.test_name = "TwoIdenticalElements",
                                 .nums = {2, 2},
                                 .target = 4,
                                 .expected = {0, 1}},
                      TwoSumCase{.test_name = "LargePositiveNegative",
                                 .nums = {1000, -1000},
                                 .target = 0,
                                 .expected = {0, 1}},
                      TwoSumCase{.test_name = "LargerArray",
                                 .nums = {1, 2, 3, 4, 5, 6},
                                 .target = 11,
                                 .expected = {4, 5}},
                      TwoSumCase{.test_name = "TwoZerosSumToZero",
                                 .nums = {0, 0},
                                 .target = 0,
                                 .expected = {0, 1}}),
    [](const ::testing::TestParamInfo<TwoSumCase> &info) {
      return info.param.test_name;
    });
